### ISSUE TITLE
ActiveRecord query with order on an included relation is readonly

### ISSUE THREAD

**Issue by jgelo** [rails/rails#25835]
It appears that queries with an order on an included relationship are causing the readonly flag to be set on the related model in Rails 5.   Using the following models:

``` ruby
class Campaign < ActiveRecord::Base
  has_many  :appeals
end

class Appeal < ActiveRecord::Base
  belongs_to :campaign
end
```

Rails 4 behavior:

``` ruby
Appeal.where(id: 1).includes(:campaign).order('campaigns.name').first.campaign.readonly?
=> false
```

Rails 5 behavior:

``` ruby
Appeal.where(id: 1).includes(:campaign).order('campaigns.name').first.campaign.readonly?
=> true
```

Some additional examples of Rails 5 behavior:

``` ruby
# ordering by a base model attribute works:
Appeal.where(id: 1).includes(:campaign).order('appeals.name').first.campaign.readonly?
=> false

# using joins instead of includes works:
Appeal.where(id: 1).joins(:campaign).order('campaigns.name').first.campaign.readonly?
=> false

# excluding the order call works:
Appeal.where(id: 1).includes(:campaign).first.campaign.readonly?
=> false

# adding a readonly call doesn't help:
Appeal.where(id: 1).includes(:campaign).order('campaigns.name').readonly(false).first.campaign.readonly?
=> true
```

Is this a regression or expected behavior in Rails 5?

**Comment by prakashmurthy** [comment]
Created an executable script to describe the behavior. The script works on `4-2-stable` and fails with master branch.

<details>
  <summary>readonly_behavior.rb</summary>

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'

  # Fail cases
  gem 'activerecord', "5.0.0" # => Fail
  # gem 'activerecord', "5.1.0.alpha" # => Fail

  # Pass case
  # gem 'activerecord', "4.2.7.rc1" # => Pass
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :campaigns, force: true do |t|
    t.string  :name
  end

  create_table :appeals, force: true do |t|
    t.string :name
    t.integer :campaign_id
  end
end

class Campaign < ActiveRecord::Base
  has_many :appeals
end

class Appeal < ActiveRecord::Base
  belongs_to :campaign
end

class BugTest < Minitest::Test
  def test_readonly_behavior
    campaign = Campaign.create(name: "First")
    Appeal.create(campaign: campaign, name: "First")
    Appeal.create(campaign: campaign, name: "Second")
    Appeal.create(campaign: campaign, name: "Third")

    assert_equal false, Appeal.where(id: 1).includes(:campaign).order('campaigns.name').first.campaign.readonly?
  end
end
```

</details>

**Comment by maclover7** [comment]
Confirming that this works on `4-2-stable`, but not on `master`.

@prakashmurthy, thanks for creating a reproduction script -- can you try and run a `git bisect`, and figure out where the current behavior was introduced?

**Comment by prakashmurthy** [comment]
@maclover7  I did try `git bisect` and ran into different error messages that I wasn't sure how to move ahead from. I have scheduled time to pair with @prathamesh-sonpatki tomorrow morning to work on git bisecting this issue; he should be able to guide me through the process. Thanks.

**Comment by prakashmurthy** [comment]
`git bisect` ends up showing https://github.com/rails/rails/commit/a975407a0a4f8bb06f8d4696de6a8b5e5a086622 as the first bad commit, but not sure that is related to the readonly behavior being here. 

Will try looking through the readonly related commits and code changes later on.

**Comment by prakashmurthy** [comment]
Went through the `git bisect` steps again, this time specifying the correct commit reference for `arel` by commit date instead of gem version, and ended up with 068160b444a5ceccd0bdc5dd81b70bc45754bcaa as the first bad commit. Seems right as it is calling `model.readonly!`

**Comment by maclover7** [comment]
@jgelo Can you check and see if #24099 fixes this issue?

**Comment by prakashmurthy** [comment]
The reproduction script works fine with https://github.com/rails/rails/pull/24099 .

**Comment by jgelo** [comment]
I can also confirm that #24099 resolves the issue.

Many thanks to all for looking into this!

**Comment by maclover7** [comment]
👍

**Comment by kamipo** [comment]
Fixed by #24099
